VAE Selector

R-N opened the pull request:
Related feature requests: 

- #3655 
- #3415 
- #3866 

Dropdown to select VAE in Settings tab. You can also use "sd_vae" quicksettings to have it on top like the checkpoint dropdown.

![image](https://user-images.githubusercontent.com/1442761/198885464-086c0634-7097-4be2-8800-0a80fc08d19d.png)

It has these options:

1. auto (default)
2. None
3. VAEs found

It will search VAEs in this order, and **priority** if same name exists will be in **reverse** order, supports subfolder:

1. /models/Stable-diffusion/*.vae.ckpt
2. /models/Stable-diffusion/*.vae.pt
3. /models/VAE/*.ckpt
4. /models/VAE/*.pt
5. the vae-path cmd line arg

It will try to load VAE in this order:

1. vae-path cmd line arg, only for the first load
2. selected VAE in the settings, fallbacks to auto if not found

If auto:

3. vae-path cmd line arg again, but also applies for non-first loads
4. .vae.pt beside the selected sd checkpoint
5. .vae.ckpt beside the selected sd checkpoint

Note: "beside" means in the same directory and with the same file name.

Issues and future improvements:

~~1. Currently, changing VAE is done by reloading the whole sd checkpoint. The reload checkpoint function has many things I don't understand whether they're also necessary for reloading VAE and how to adjust them for reloading VAE. I'll try to improve this later when I have the time.~~
~~2. Due to issue 1, reloading VAE bypasses caching.~~
~~2. Due to issue 1, caching now works by checkpoint-vae pair as key so it's inefficient. Though if you're not changing VAE, it's pretty much the same as before (just a bit larger key).~~
~~3. VAE loaded through cmd line arg isn't saved nor reflected in the settings dropdown. This means the dropdown can show totally different VAE from the actually loaded one. I don't know how to set it through code. Simply setting it will break it. I'm not sure if saving it is a good idea either.~~
4. The currently loaded VAE from auto isn't reflected in the settings dropdown nor displayed anywhere (other than the output log). The dropdown stays "auto". I'm not sure what to do about this.
5. The base VAE is stored in RAM when you load a separate VAE (for preparation when you select None). I guess this should be a setting (whether you want to store it in RAM, in file, or not at all so selecting None will require reloading checkpoint).

R-N committed:
Settings to select VAE

R-N committed:
Forgot to add this folder

R-N committed:
Change VAE search order and thus priority

R-N committed:
Fix typo from previous commit

R-N committed:
Checkpoint cache by combination key of checkpoint and vae

Ehplodor commented:
For my personal interest, is there any fair comparison between VAEs ? I couldn't find any through google nor in this repository's issues or discussions. Thanks in advance.

R-N commented:
> For my personal interest, is there any fair comparison between VAEs ? I couldn't find any through google nor in this repository's issues or discussions. Thanks in advance.

[This](https://huggingface.co/stabilityai/sd-vae-ft-mse-original) is the only one I've come across.

R-N commented:
> The refresh button spans across a large chunk of the quick settings bar. Pic below. ![image](https://user-images.githubusercontent.com/31809330/199004013-72e3216a-69fd-4fe2-bfce-ae4387a94b9c.png)
> ...

Yeah it seemed a bit longer on my screenshot too but I thought it was fine. Turns out it stretches out to full width

Thanks for the fix.

R-N commented:
Also I'll try to improve the loading later when I have the time. I think we can reload VAE without reloading the full checkpoint. Maybe someone can give hints on these, specially the first one.

<details>
<summary>reload_model_weights</summary>

```
# What does this do? 
# Is it needed for VAE reload?
if sd_model.sd_checkpoint_info.config != checkpoint_info.config or should_hijack_inpainting(checkpoint_info) != should_hijack_inpainting(sd_model.sd_checkpoint_info):
    checkpoints_loaded.clear()
    load_model(checkpoint_info)
    return shared.sd_model

# Send model to CPU, or everything if lowvram
if shared.cmd_opts.lowvram or shared.cmd_opts.medvram:
    lowvram.send_everything_to_cpu()
else:
    sd_model.to(devices.cpu)

# What does hijack do?
# I think this should also be called when reloading VAE, at least for consistency?
sd_hijack.model_hijack.undo_hijack(sd_model)

load_model_weights(sd_model, checkpoint_info)

# Related to previous one
sd_hijack.model_hijack.hijack(sd_model)
# I think this callback should also be called when reloading VAE?
script_callbacks.model_loaded_callback(sd_model)

# Return models to appropriate device
if not shared.cmd_opts.lowvram and not shared.cmd_opts.medvram:
    sd_model.to(devices.device)
```
</details>

However to set VAE to None, it means returning the VAE weights to the checkpoint's default. I think for this one we'd still need to reload the checkpoint, or cache the checkpoint's unaltered VAE weights. How big can VAE be in memory? Is caching in RAM okay? Maybe this should be an option/setting?

R-N committed:
Fix VAE refresh button stretching out

From https://github.com/AUTOMATIC1111/stable-diffusion-webui/pull/3986#issuecomment-1296990601

sedetweiler commented:
> For my personal interest, is there any fair comparison between VAEs ? I couldn't find any through google nor in this repository's issues or discussions. Thanks in advance.

This video is also helpful for seeing the differences: https://youtu.be/QFFQi5Jd4I4

Ehplodor commented:
> > For my personal interest, is there any fair comparison between VAEs ? I couldn't find any through google nor in this repository's issues or discussions. Thanks in advance.
> 
> This video is also helpful for seeing the differences: https://youtu.be/QFFQi5Jd4I4

This ID a very good presentation, indeed. Thanks

AUTOMATIC1111 commented:
># What does this do? 
># Is it needed for VAE reload?
If a model has a config file next to it, reloading just weights is not possible, you have to make the model from scratch.

># I think this should also be called when reloading VAE, at least for consistency?
Hijack is removed and re-added because it changes layer names for CLIP model, so when reloading weights if you don't remove it, CLIP is juts not loaded.

># Why not return model back to device when low vram?
In lowvram mode only one part of model is in VRAM, the rest of it stays in CPU memory until it is needed. And when a part of model is about to be used, it is automatically sent to GPU. So you do not return entire model to VRAM because that's how lowvram mode works.

R-N commented:
Thank you. I still have some questions though

> If a model has a config file next to it, reloading just weights is not possible, you have to make the model from scratch.

Does the config affect the VAE part too?

And what about the inpainting part?

R-N committed:
Merge branch 'master' into vae-picker

R-N commented:
Nvm, just checked the sd_models.py. 

So the config is needed to instantiate the base model. Then checkpoint weights are loaded into it, then VAE. Thus if config is different, model needs to be reinstantiated. However if config is the same there's no need to reinstantiate. This should apply to VAE loading too. 

Meanwhile the inpainting part changes config, and it's hardcoded. So normal and inpainting models are instantiated with different config despite having the same or no config file.

R-N committed:
Reload VAE without reloading sd checkpoint

R-N commented:
Tests of two rounds of changing VAEs for nai model with below order. Same prompt, seed, and other settings. No xformers. Images generated using same VAE should be the same.

Prompt: 
```
ganyu \(genshin impact\),
Steps: 20, Sampler: Euler a, CFG scale: 7, Seed: 1, Size: 512x512, Model hash: 925997e9, Eta: 0.0022321428125, Clip skip: 2, ENSD: 31337
```

Order:
1. animevae
2. wd_kl-f8-anime2
3. sd_1.5_mse
4. None
5. animevae
6. wd_kl-f8-anime2
7. sd_1.5_mse
8. None

<details><summary>First round (for comparison between VAEs used)</summary>

animevae:
![00007-1-ganyu _(genshin impact___](https://user-images.githubusercontent.com/1442761/199409529-cacbdfd3-2c41-418e-a3a6-2c8129c6111e.png)
wd_kl-f8-anime2:
![00008-1-ganyu _(genshin impact___](https://user-images.githubusercontent.com/1442761/199409628-0640dd6f-8754-42d5-ae5a-c5efadb41b37.png)
sd_1.5_mse:
![00011-1-ganyu _(genshin impact___](https://user-images.githubusercontent.com/1442761/199409719-4481a8f1-01cd-45ef-b7d1-0a2231b1e578.png)
None:
![00012-1-ganyu _(genshin impact___](https://user-images.githubusercontent.com/1442761/199409789-1baaa758-da24-4bfc-9105-06905123b708.png)
</details>

For comparison between rounds (grouped by VAEs, each having two images which should be the same):
<details><summary>animevae</summary>

![00007-1-ganyu _(genshin impact___](https://user-images.githubusercontent.com/1442761/199409529-cacbdfd3-2c41-418e-a3a6-2c8129c6111e.png) ![00013-1-ganyu _(genshin impact___](https://user-images.githubusercontent.com/1442761/199409548-d492c04d-a618-477e-b973-1b46f29be4e7.png)
</details>

<details><summary>wd_kl-f8-anime2</summary>

![00008-1-ganyu _(genshin impact___](https://user-images.githubusercontent.com/1442761/199409628-0640dd6f-8754-42d5-ae5a-c5efadb41b37.png) ![00014-1-ganyu _(genshin impact___](https://user-images.githubusercontent.com/1442761/199409649-651a192e-1f0c-44d3-b47d-a7ca49dc1476.png)
</details>

<details><summary>sd_1.5_mse</summary>

![00011-1-ganyu _(genshin impact___](https://user-images.githubusercontent.com/1442761/199409719-4481a8f1-01cd-45ef-b7d1-0a2231b1e578.png) ![00015-1-ganyu _(genshin impact___](https://user-images.githubusercontent.com/1442761/199409730-15fafd81-371b-498a-9bdc-90d69c32cad0.png)
</details>

<details><summary>None</summary>

![00012-1-ganyu _(genshin impact___](https://user-images.githubusercontent.com/1442761/199409789-1baaa758-da24-4bfc-9105-06905123b708.png) ![00016-1-ganyu _(genshin impact___](https://user-images.githubusercontent.com/1442761/199409811-402532bd-c5e4-46b8-b150-066e394f03fe.png)
</details>

R-N committed:
Save VAE provided by cmd_opts.vae_path

Miraculix200 commented:
I think the --vae-path option doesn't do what it's supposed to do? When adding a path, it loads the .ckpt files from the vae models path into the "Stable Diffusion checkpoints" dropdown list, and no .pt or .vae.pt files from this folder are visible in the SD VAE dropdown

younyokel commented:
I don't know if I should post this here. But is it possible to make a checkbox in the Settings so that custom VAE is ignored if there is a ".vae.pt" file found in the directory with the same name as the selected model? For example, I would like to open ft-mse-original VAE for everything EXCEPT the NovelAI model.

R-N commented:
> I don't know if I can post this here, but is it possible to make a checkbox in Settings to ignore a custom VAE if a ".vae.pt" file with the same name as the selected model is found in the directory, as before? For example, I would like to open ft-mse-original VAE for everything **except** the NovelAI model.

So, if I understand correctly, you want to be able to prioritize the "similar" VAEs, so that the selected VAEs become a default fallback instead? Is this like #3655? (I mentioned it but it's not exactly what this PR does, just related. Maybe it should be reopened) 
Is a checkbox in the settings alright? Should it be prioritized over --vae-path command line option too?

aliencaocao commented:
Hi for some reason the quick setting for vae selector isn't working for me - I only see the quick selector for model checkpoint and hypternetwork. Do you have any idea why?
my quick settings is `sd_model_checkpoint, sd_hypernetwork,setting_sd_vae`

Kronzky commented:
Would it be possible to make the vae-path work similar to the ckpt-dir argument, i.e. you only specify the path to the *folder*, and can then select the desired file in the drop-down menu?
Right now, it seems, you can only specify the path to the actual VAE *file* it should start with, but when you then go to the drop-down menu it doesn't know anything about any other VAE files that may be in the same folder.

Woisek commented:
Is there any possibility, that the auto function of SD VAE option not respect safetensors or is working at all?
Looking at the list ...

1.     /models/Stable-diffusion/*.vae.ckpt
2.     /models/Stable-diffusion/*.vae.pt
3.     /models/VAE/*.ckpt
4.     /models/VAE/*.pt
5.     the vae-path cmd line arg

I can't see if .safetensors are also auto loaded. But probably not, I guess.
On the other hand, for example, I have vae-ft-mse-840000-ema-pruned.ckpt and vae-ft-mse-840000-ema-pruned.safetensors
The Model Art & Eros (https://civitai.com/models/3950/art-and-eros-aeros-a-tribute-to-beauty) as pruned version needs a VAE. But selecting the SD VAE option to "auto" does nothing, images coming out funky. Setting it to vae-ft-mse-840000-ema-pruned.* and everything is OK.

So, my question is: Does this "auto" function really work? From my point of view, no.

Any ideas or suggestions?

R-N commented:
@Woisek  This PR was made before safetensors were a thing. Also [Automatic1111 has rewritten the VAE selector, and the code is now nothing alike](https://github.com/AUTOMATIC1111/stable-diffusion-webui/pull/4843#issuecomment-1383238565). Consider opening a new issue.

Woisek commented:
@R-N 
Thanks for this info. But much more I would like to know about the "auto" thing. Like I think, this doesn't work (anymore), at least that's what my experience is. Therefore it would be nice to know if this is a bug or not, will it be fixed or not, is it something on my end I'm doing wrong ...

Questions over questions ... :)

R-N commented:
@Woisek I don't know. I haven't used the auto since ages. I usually just select VAE from the dropdown because it's now so fast to switch VAE. 

Is your model checkpoint name also vae-ft-mse-840000-ema-pruned? Auto loads VAE with the same name from the same location (only different file extension).

Woisek commented:
@R-N 
>Is your model checkpoint name also vae-ft-mse-840000-ema-pruned?

No, and I guess that wouldn't make any sense, since some Checkpoints come with special inpainting versions and same name would not work, except renameing the VAE. But this is also nonsense, to copy the same VAE all over the place.

>Auto loads VAE with the same name as the selected model from the same directory (only different file extension).

If that's the case, the auto function is IMHO useless. When I have to copy the same VAE over and over again, just with different name, what's the point in "auto", as the VAE with the same name would be loaded anyway (as far as I understood this all).

The use of "auto", for me, would be, that there is "one VAE to rule them all" ... :) I mean, if one specific VAE is present, the auto function detects it and applies/activates it wenn a model is selected, that needs it. That's my understanding auf "auto".

>Also the model you mentioned has a version that doesn't require loading a separate VAE.

Yes, I know, that was just as an example. I just don't know if it's a disadvantage in regads to memory/performance/output when models with VAE included are used with a (maybe different) VAE selected.

R-N commented:
>except renameing the VAE. 
Yes that's the idea. Auto is there to support old behavior where WebUI automatically identify and load VAE of a checkpoint indicated by having the same name.

>But this is also nonsense, to copy the same VAE all over the place.
Yeah that's why the VAE selector was made.


> what's the point in "auto", as the VAE with the same name would be loaded anyway (as far as I understood this all).
No, if you don't select Auto, WebUI will load the VAE selected in the dropdown (or from the --vae-path cmd arg, if you set it). Not the one with same name.

The point is as I mentioned before, to support old behavior.

> if one specific VAE is present, the auto function detects it and applies/activates it wenn a model is selected, that needs it. That's my understanding auf "auto".
There is no way to know which model "needs" which VAE.

Woisek commented:
OK, so the bottom line of all this is, that VAE will die out in the foreseeable future ... ?

R-N commented:
> OK, so the bottom line of all this is, that VAE will die out in the foreseeable future ... ?

Separate VAE file? I guess so. 

The feature will likely stay though

Woisek commented:
Yes, I meant the separate VAE.

enzyme69 commented:
Hello just info, as for today, this VAE dropdown is in **settings > stable diffusion**

Thanks discord~

RykeWollf commented:
Hi there, I'm a complete noob at all of this so I'm surprised I made it this far. I don't know if I'm allowed to even ask this, but how can I integrate this on my SD? 

Thank you!

Ehplodor commented:
@RykeWollf look for quicksetings and : 
<img width="213" alt="image" src="https://user-images.githubusercontent.com/80175881/221873861-48e7a227-3dfd-486c-9b86-cf627f20e451.png">

RykeWollf commented:
> @RykeWollf look for quicksetings and : <img alt="image" width="213" src="https://user-images.githubusercontent.com/80175881/221873861-48e7a227-3dfd-486c-9b86-cf627f20e451.png">

oh this is such a time saver! Thank you @Ehplodor ! Is there anything similar for clip skip, loras and hypernetworks?
